sys/kern/kmem.c

```diff
@@ -28,8 +30,26 @@ static void kick_swapper(void) {
 vaddr_t kva_alloc(size_t size) {
   assert(page_aligned_p(size));
   vmem_addr_t start;
-  if (vmem_alloc(kvspace, size, &start, M_NOGROW))
-    return 0;
+
+  bool restarted = false;
+
+retry:
+  if (vmem_alloc(kvspace, size, &start, M_NOGROW)) {
+    if (restarted)
+      return 0;
+
+    WITH_MTX_LOCK (&maxkvaddr_lock) {
+      vaddr_t old = maxkvaddr;
+      maxkvaddr = pmap_growkernel(maxkvaddr + size);
+      klog("%s: increase kernel end %08lx -> %08lx", __func__, old, maxkvaddr);
+
+      if (vmem_add(kvspace, old, maxkvaddr - old))
+        return 0;
+    }
+
+    restarted = true;
+    goto retry;
+  }
```

Comment:
Consider two threads: A and B. Both threads concurrently call `kva_alloc()` and the first call to `vmem_alloc()` fails in both threads. Then, thread A grows `kvspace` and after that, thread B grows `kvspace`. We end up with `kvspace` grown by too much, as growing it only once would be sufficient for both threads.
I think we could fix this by checking if the value of `maxkvaddr` changed between the first call to `vmem_alloc()` and acquiring `maxkvaddr_lock`:
```c
retry:
  vaddr_t old = maxkvaddr;
  if (vmem_alloc(kvspace, size, &start, M_NOGROW)) {
    if (restarted)
      return 0;

    WITH_MTX_LOCK (&maxkvaddr_lock) {
      if (maxkvaddr > old) {
        mtx_unlock(&maxkvaddr_lock);
        goto retry;
      }
      ...
    }

    restarted = true;
    goto retry;
  }

```

Reply:
done